miden-tx/src/testing/tx_context.rs

```diff
@@ -197,21 +216,385 @@ impl TransactionContextBuilder {
+                # ---------------------------------------------------------------------------------
+                push.{recipient0}
+                push.{PUBLIC_NOTE}
+                push.{aux0}
+                push.{tag0}
+                call.{ACCOUNT_CREATE_NOTE_MAST_ROOT}
+
+                push.{asset0} movup.4
+                call.{ACCOUNT_ADD_ASSET_TO_NOTE_MAST_ROOT}
+                dropw dropw dropw
+
+                # NOTE 1
+                # ---------------------------------------------------------------------------------
+                push.{recipient1}
+                push.{PUBLIC_NOTE}
+                push.{aux1}
+                push.{tag1}
+                call.{ACCOUNT_CREATE_NOTE_MAST_ROOT}
+
+                push.{asset1} movup.4
+                call.{ACCOUNT_ADD_ASSET_TO_NOTE_MAST_ROOT}
+                dropw dropw dropw
+            end
+            ",
+            PUBLIC_NOTE = NoteType::Public as u8,
+            recipient0 = prepare_word(&output0.recipient().digest()),
+            aux0 = output0.metadata().aux(),
+            tag0 = output0.metadata().tag(),
+            asset0 = prepare_assets(output0.assets())[0],
+            recipient1 = prepare_word(&output1.recipient().digest()),
+            aux1 = output1.metadata().aux(),
+            tag1 = output1.metadata().tag(),
+            asset1 = prepare_assets(output1.assets())[0],
+        );
+
+        NoteBuilder::new(sender, ChaCha20Rng::from_seed(self.rng.gen()))
+            .note_inputs(inputs)
+            .unwrap()
+            .add_assets([asset])
+            .code(code)
+            .build(&self.assembler)
+            .unwrap()
+    }
+
+    fn input_note_transfer(
+        &mut self,
+        sender: AccountId,
+        assets: impl IntoIterator<Item = Asset>,
+    ) -> Note {
+        let code = "
+            use.miden::note
+            use.miden::contracts::wallets::basic->wallet
+
+            begin
+                # read the assets to memory
+                push.0 exec.note::get_assets
+                # => [num_assets, dest_ptr]
+
+                # assert the number of assets is 3
+                push.3 assert_eq
+                # => [dest_ptr]
+
+                # add the first asset to the vault
+                padw dup.4 mem_loadw call.wallet::receive_asset dropw
+                # => [dest_ptr]
+
+                # add the second asset to the vault
+                push.1 add padw dup.4 mem_loadw call.wallet::receive_asset dropw
+                # => [dest_ptr+1]
+
+                # add the third asset to the vault
+                push.1 add padw movup.4 mem_loadw call.wallet::receive_asset dropw
+                # => []
+            end
+        ";
+
+        NoteBuilder::new(sender, ChaCha20Rng::from_seed(self.rng.gen()))
+            .add_assets(assets)
+            .code(code)
+            .build(&self.assembler)
+            .unwrap()
+    }
+
+    pub fn with_mock_notes_too_few_input(mut self) -> Self {
```

Comment:
Just like the accounts, I broke up the scenarios into different methods. This makes it easier to modify the scenarios, since the function body contains only the code relevant to the scenario at hand. I'm going to stop here, since at this stage it is easy to add a `input_note_transfer_delayed_authentication` of sorts, which is what I need for testing, but I think these methods need better names (they are completely cryptic at the moment), some documentation, and some could even be removed because they are not used more than once.